api/jobs/batch.py

```diff
@@ -0,0 +1,182 @@
+            if not_matched:
+                not_matched_conts.append(c)
+            elif ambiguous:
+                ambiguous_conts.append(c)
+            else:
+                # Create input map of file refs
+                inputs = {}
+                for input_name, files in suggestions.iteritems():
+                    inputs[input_name] = {'type': container_type, 'id': str(c['_id']), 'name': files[0]}
+                c['inputs'] = inputs
+                matched_conts.append(c)
+        else:
+            not_matched.append(c)
+    return {
+        'matched': matched_conts,
+        'not_matched': not_matched_conts,
+        'ambiguous': ambiguous_conts
+    }
+
+def insert(batch_proposal):
+    """
+    Simple database insert given a batch proposal.
+    """
+    return config.db.batch.insert(batch_proposal)
+
+def update(batch_id, payload):
+    """
+    Updates a batch job, being mindful of state flow.
+    """
+
+    bid = bson.ObjectId(batch_id)
+    query = {'_id': bid}
+    if payload.get('state'):
+        # Require that the batch job has the previous state
+        query['state'] = BATCH_JOB_TRANSITIONS[payload.get('state')]
+    result = config.db.batch.update_one({'_id': bid}, {'$set': payload})
+    if result.modified_count != 1:
+        raise Exception('Batch job not updated')
+
+def run(batch_job):
+    """
+    Creates jobs from proposed inputs, returns jobs enqueued.
+    """
+
+    proposed_inputs = batch_job.get('proposed_inputs', [])
+    gear_name = batch_job.get('gear')
+    config_ = batch_job.get('config')
+    origin = batch_job.get('origin')
+
+    jobs = []
+    job_ids = []
+    for inputs in proposed_inputs:
+        for input_name, fr in inputs.iteritems():
+            inputs[input_name] = create_filereference_from_dictionary(fr)
+        # TODO support analysis gears (will have to create analyses here)
+        destination = create_containerreference_from_filereference(inputs[inputs.keys()[0]])
+        job = Job(gear_name, inputs, destination=destination, tags=['batch'], config_=config_, origin=origin)
+        job_id = job.insert()
+        jobs.append(job)
+        job_ids.append(job_id)
+
+    update(batch_job['_id'], {'state': 'launched', 'jobs': job_ids})
+    return jobs
+
+def cancel(batch_job):
+    """
+    Cancels all pending jobs, returns number of jobs cancelled.
+    """
+
+    pending_jobs = config.db.jobs.find({'state': 'pending', '_id': {'$in': batch_job.get('jobs')}})
+    cancelled_jobs = 0
+    for j in pending_jobs:
+        job = Job.load(j)
+        try:
+            Queue.mutate(job, {'state': 'cancelled'})
+            cancelled_jobs += 1
+        except Exception: # pylint: disable=broad-except
+            # if the cancellation fails, move on to next job
+            continue
+
+    update(batch_job['_id'], {'state': 'cancelled'})
+    return cancelled_jobs
+
+
+def get_stats():
+    """
+    Return the number of jobs by state.
+    """
```

Comment:
I'm not sure if this is useful to you @dpuccetti but I'll add it before merging. It's a simple mongo aggregation.